dcd/usb_phydcd.c

```diff
@@ -100,154 +111,476 @@ usb_phydcd_status_t USB_PHYDCD_Control(usb_phydcd_handle handle, usb_phydcd_cont
+}
+
+static usb_phydcd_dev_status_t dcdSecondaryDetection(usb_phydcd_state_struct_t *dcd)
+{
+    usb_phydcd_dev_status_t ret = dcd->dcdDetectState;
+
+    /* Secondary detection after set time*/
+    if ((dcd->hwTick - dcd->startTime) >= USB_DCD_SECONDARY_DETECTION_TIME) {
+        /* DM line state output of the charger detector (DM_STATE)
+         * 		DM_STATE
+         */
+        dcd->detectResult = (usbAnalogChargerDetectStatGet(dcd, USB_ANALOG_CHRG_DETECT_STAT_DM_STATE_MASK))
+                                ? (uint8_t)kUSB_DcdDCP
+                                : (uint8_t)kUSB_DcdCDP;
+
+        /* Setting this bit can enable 1.5 kΩ pull-up resister on DP.(UTMI_TESTSTART)
+         * 		Setting this bit can enable 1.5 kΩ pull-up resister on DP.
+         *
+         * 		This bit can only be used as DCD detection, while it must be cleared in normal function.
+         */
+        usbLoopbackRegSet(dcd, CLEAR, USB_ANALOG_LOOPBACK_UTMI_TESTSTART_MASK);
+
+        /* (HSTPULLDOWN)
+         * Set bit 3 to 1 to pull down 15-KOhm on USB_DP line.
+         * Set bit 2 to 1 to pull down 15-KOhm on USB_DM line.
+         * Clear to 0 to disable.
+         */
+        usbPhyDebugRegSet(dcd, CLEAR, USB_DCD_SECONDARY_DETECTION_PULL_DOWN_CONFIG);
+
+        /* Gate Test Clocks. (CLKGATE)
+         * 		Clear to 0 for running clocks.
+         * 		Set to 1 to gate clocks. Set this to save power while the USB is not actively being used
+         */
+        usbPhyDebugRegSet(dcd, SET, USBPHY_DEBUG_CLR_CLKGATE_MASK);
+
+        /* Set new state start time */
+        dcd->startTime = dcd->hwTick;
+
+        /* Change state machine to detection finished */
+        ret = kUSB_DCDDectionFinished;
+    }
+
+    return ret;
+}
+
+static usb_phydcd_dev_status_t dcdDetectionFinished(usb_phydcd_state_struct_t *dcd)
+{
+    static uint32_t run_Times = 0;
+    usb_phydcd_dev_status_t next_step;
+
+    if (run_Times == 0U) {
+        // LOG_INFO("dcdDetectionFinished, detectResult: %u", dcd->detectResult);
+        dcd->detectResult = (uint8_t)kUSB_DcdUnknownType;
+        next_step         = kUSB_DCDDetectWait;
+    }
+    else {
+        next_step = kUSB_DCDDetectIdle;
+    }
+
+    (void)dcd->dcdCallback(dcd->dcdCallbackParam, dcd->detectResult, (void *)&dcd->detectResult);
+
+    run_Times++;
+
+    /* Set new state start time */
+    dcd->startTime = dcd->hwTick;
+    /* Change state machine to Idle */
+    return next_step;
+}
+
+static usb_phydcd_dev_status_t dcdDetectionWait(usb_phydcd_state_struct_t *dcd)
+{
+    if ((dcd->hwTick - dcd->startTime) >= 1500u) {
+        /* Set new state start time */
+        dcd->startTime = dcd->hwTick;
+        /* Change state machine to Idle */
+        return kUSB_DCDDetectInit;
+    }
+    return kUSB_DCDDetectWait;
+}
+
+usb_phydcd_status_t USB_PHYDCD_TimerIsrFunction(usb_phydcd_handle handle, const uint64_t tick)
+{
+    usb_phydcd_status_t ret = ((NULL == handle) ? kStatus_phydcd_Error : kStatus_phydcd_Success);
```

Comment:
Use early return pattern here to just leave this function if the handle is `NULL`.